src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs

```diff
@@ -55,123 +53,192 @@ from typeHandled in messagesBeingHandled
+    readonly Dictionary<Type, List<IMessageHandlerFactory>> messageHandlerFactories = [];
+    readonly HashSet<HandlerAndMessage> deduplicationSet = [];
+    static readonly Type IHandleMessagesType = typeof(IHandleMessages<>);
+    static readonly ILog Log = LogManager.GetLogger<MessageHandlerRegistry>();
 
-        return Expression.Lambda<Func<object, object, IMessageHandlerContext, Task>>(body, target, messageParam, contextParam).CompileFast();
+    readonly record struct HandlerAndMessage(Type HandlerType, Type MessageType, bool IsTimeoutHandler)
+    {
+        public static HandlerAndMessage New<THandler, TMessage>(bool isTimeoutHandler = false) =>
+            new(HandlerType: typeof(THandler), MessageType: typeof(TMessage), IsTimeoutHandler: isTimeoutHandler);
     }
 
-    static Type[] GetMessageTypesBeingHandledBy(Type type)
+    interface IMessageHandlerFactory
     {
-        return (from t in type.GetInterfaces()
-                where t.IsGenericType
-                let potentialMessageType = t.GetGenericArguments()[0]
-                where
-                typeof(IHandleMessages<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t) ||
-                typeof(IHandleTimeouts<>).MakeGenericType(potentialMessageType).IsAssignableFrom(t)
-                select potentialMessageType)
-            .Distinct()
-            .ToArray();
+        Type MessageType { get; }
+        MessageHandler Create();
     }
 
-    static void ValidateHandlerType(Type handlerType)
+    sealed class TimeoutHandlerFactory<THandler, TMessage> : IMessageHandlerFactory
+        where THandler : class
     {
-        var propertyTypes = handlerType.GetProperties().Select(p => p.PropertyType).ToList();
-        var ctorArguments = handlerType.GetConstructors()
-            .SelectMany(ctor => ctor.GetParameters().Select(p => p.ParameterType))
-            .ToList();
+        public Type MessageType { get; } = typeof(TMessage);
 
-        var dependencies = propertyTypes.Concat(ctorArguments).ToList();
+        public MessageHandler Create() =>
+            new MessageHandlerInvoker<IHandleTimeouts<TMessage>, TMessage>(
+                static provider => handlerFactory(provider, []),
+                static (handler, message, handlerContext) => handler.Timeout(message, handlerContext),
+                isTimeoutHandler: true)
+            {
+                HandlerType = typeof(THandler)
+            };
 
-        if (dependencies.Any(t => typeof(IMessageSession).IsAssignableFrom(t)))
-        {
-            throw new Exception($"Interfaces IMessageSession or IEndpointInstance should not be resolved from the container to enable sending or publishing messages from within sagas or message handlers. Instead, use the context parameter on the {handlerType.Name}.Handle method to send or publish messages.");
-        }
-    }
+        static readonly ObjectFactory<THandler> factory = ActivatorUtilities.CreateFactory<THandler>([]);
```

Comment:
The naming of these two factories is a little trippy but I'm not sure (yet) what I'd want to do about it.